### PR TITLE
Smithsonian Transformer

### PR DESCRIPTION
**Pull request by dannymandel**
Initial support for the Smithsonian transformer.  Still need to hook up the controlled vocabulary support

### CODE REVIEW

File: isamples_metadata/GEOMETransformer.py [hunk 1]
```diff
@@ -30,10 +30,6 @@ def __init__(self, source_record: typing.Dict):
                     GEOMEChildTransformer(source_record, child_record)
                 )
```
Comment (dannymandel) [inline isamples_metadata/GEOMETransformer.py:32]: Since I was creating a new one of these, it seemed like a good time to revisit previous implementations and clean up mistakes.  All the changes in files other than Smithsonian are simple refactorings to clean the code up.

File: isamples_metadata/OpenContextTransformer.py [hunk 2]
```diff
@@ -100,29 +100,16 @@ def categories_mappers(cls) -> typing.List[AbstractCategoryMapper]:
 
 
 class OpenContextTransformer(Transformer):
```
Comment (dannymandel) [inline isamples_metadata/OpenContextTransformer.py:102]: Cleanup

File: isamples_metadata/SESARTransformer.py [hunk 3]
```diff
@@ -225,10 +225,6 @@ def categories_mappers(cls) -> typing.List[AbstractCategoryMapper]:
 class SESARTransformer(Transformer):
```
Comment (dannymandel) [inline isamples_metadata/SESARTransformer.py:225]: Cleanup

File: isamples_metadata/Transformer.py [hunk 4]
```diff
@@ -21,13 +29,25 @@ def _transform_key_to_label(
         if label is None:
             label = key
         value = source_dict.get(key)
-        if value is not None:
+        if value is not None and len(value) > 0:
```
Comment (dannymandel) [inline isamples_metadata/Transformer.py:32]: One major difference is that the Smithsonian records (at least in my tests) had empty strings.  I should probably fix this on the frontend things import to not include empty string values.  That would make our lives easier downstream here…

File: isamples_metadata/Transformer.py [hunk 5]
```diff
@@ -11,6 +11,14 @@ class Transformer(ABC):
...
     DESCRIPTION_SEPARATOR = " | "
 
+    N2T_PREFIX = "https://n2t.net/"
+
+    N2T_ARK_PREFIX = f"{N2T_PREFIX}ark:/"
+
+    N2T_NO_HTTPS_PREFIX = "http://n2t.net/"
+
+    N2T_ARK_NO_HTTPS_PREFIX = f"{N2T_NO_HTTPS_PREFIX}ark:/"
+
```
Comment (dannymandel) [inline isamples_metadata/Transformer.py:21]: Unfortunately the Smithsonian export doesn't use `https` in their n2t URLs.  =\

File: isamples_metadata/Transformer.py [hunk 6]
```diff
@@ -87,16 +107,6 @@ def id_string(self) -> typing.AnyStr:
     def sample_identifier_string(self) -> typing.AnyStr:
         pass
```
Comment (dannymandel) [inline isamples_metadata/Transformer.py:109]: Cleanup

File: isamples_metadata/Transformer.py [hunk 7]
```diff
@@ -21,13 +29,25 @@ def _transform_key_to_label(
...
+        if len(month) > 0:
+            result_time_pieces.append(month.zfill(2))
+        if len(day) > 0:
+            result_time_pieces.append(day.zfill(2))
+        return "-".join(result_time_pieces)
+
     def __init__(self, source_record: typing.Dict):
         self.source_record = source_record
 
-    @abstractmethod
```
Comment (dannymandel) [inline isamples_metadata/Transformer.py:30]: Cleanup

File: tests/test_isamples_metadata.py [hunk 8]
```diff
@@ -85,6 +88,10 @@ def test_geome_child_dicts_equal(geome_source_path, isamples_path):
     (
         "../examples/OpenContext/raw/ark-28722-k2m61xj9b.json",
         "../examples/OpenContext/test/ark-28722-k2m61xj9b-test.json"
+    ),
+    (
+        "../examples/OpenContext/raw/ark-28722-k2qj7np9g.json",
+        "../examples/OpenContext/test/ark-28722-k2m61xj9b-test.json"
```
Comment (dannymandel) [inline tests/test_isamples_metadata.py:94]: Unrelated cleanup -- I had this in my local copy and apparently forgot to check it in when working on OpenContext

**Review by datadavev — APPROVED** [review]
Looks good

File: isamples_metadata/isamplesfasttext.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+from fasttext.FastText import _FastText
+
+MODEL_FILENAME = "./sampledFeature.bin"
+THIS_PATH = os.path.dirname(os.path.abspath(__file__))
+MODEL_PATH = os.path.join(THIS_PATH, MODEL_FILENAME)
+MODEL = None
+RESULT_PREFIX = "__label__"
+
+
+def _model() -> _FastText:
```
Comment (dannymandel) [inline isamples_metadata/isamplesfasttext.py:17]: @datadavev this isn't thread-safe as written.  Given our current usage I think it's probably ok, but I was wondering if you have a pointer at a python idiomatic way of handling something like this.

File: isamples_metadata/isamplesfasttext.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import fasttext
+import os.path
+import typing
+import re
+
+from fasttext.FastText import _FastText
+
+MODEL_FILENAME = "./sampledFeature.bin"
+THIS_PATH = os.path.dirname(os.path.abspath(__file__))
+MODEL_PATH = os.path.join(THIS_PATH, MODEL_FILENAME)
```
Comment (dannymandel) [inline isamples_metadata/isamplesfasttext.py:12]: @datadavev I found something like this in some `isb` code, so I'm assuming it's a good way to do it.  Let me know if there's a better way.

File: isamples_metadata/isamplesfasttext.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    return MODEL
+
+
+def predict_sampled_feature(context: typing.List[typing.AnyStr]) -> typing.AnyStr:
+    """
+    Invoke the pre-trained fasttext model to predict the sampledFeature label for the specified string inputs
+    :param context: The unprocessed inputs to the model, taken from a source record
+    :return: String label to be used as the sampled feature for the source record
+    """
+
```
Comment (dannymandel) [inline isamples_metadata/isamplesfasttext.py:30]: @qgan7125 How does this look to you?  I tried to repackage your module given what we have at our disposal when we'll invoke it, then reuse the same processing logic.  Do this look right to you?  It "works" but it's difficult to say exactly whether it's correct or not.

Comment (qgan7125) [reply isamples_metadata/isamplesfasttext.py:30]: Yeah, I think it's absolutely correct! I did want to ask you whether I need to write some interfaces as same as your current work this morning but I forgot to ask.

File: isamples_metadata/isamplesfasttext.py (hunk 9 above)
Comment (datadavev) [reply isamples_metadata/isamplesfasttext.py:17]: Perhaps create a class that wraps the fastText transformer to include the configuration properties as parameters. Then in FastAPI app, create an instance of the class during startup and shared as needed

File: examples/smithonsonian/DwC raw/DwC_occurrence_10.csv [hunk 12]
```diff
@@ -1,21 +1,21 @@
...
+http://n2t.net/ark:/65665/3000768ed-fc68-446b-aa6d-f8bd437c796a	PhysicalObject		http://grbio.org/cool/142r-0w94	US	Botany	MaterialSample	AE0OW35	http://n2t.net/ark:/65665/3000768ed-fc68-446b-aa6d-f8bd437c796a	12758488		2403	Tibet-MacArthur, J. Wen, Z. Nie, L. Xie, Y. Niu, G. Li, F. Yang, S. Lutz & J. Van De Veire	1				in collection					168	168	2009	6	17	17 Jun 2009					Asia-Temperate, China, Yunnan, Deqin Xian	Asia-Temperate				China	Yunnan	Deqin Xian	Baima Shan		4350.0								28.3419	99.0661								Rhododendron hippophaeoides	Plantae, Dicotyledonae, Ericales, Ericaceae	Plantae		Dicotyledonae	Ericales	Ericaceae	Rhododendron		hippophaeoides			Balf. f. & W.W. Sm.	Tissue & Parts; Leaf; Silica Gel				tissue
```
Comment (dannymandel) [inline examples/smithonsonian/DwC raw/DwC_occurrence_10.csv:21]: Regenerated this file using pandas to glue the data from multiple Darwin Core files into one csv.
